tests/jerry/regression-test-issue-2478.js

```diff
@@ -0,0 +1,19 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+function fn(x) {
```

Comment:
It is not needed to trigger the lexer issue, when it is still present. But when the issue is fixed, this argument declares `x` for the `switch (x)`.

Reply:
Yep, it is required -- In addition to what Akos mentioned regarding `x`, a return statement outside of a function is a syntax error.

Reply:
True, it is needed because of return statement. Would this single line trigger the issue?

`new Function("switch(1) { default: { return }}")`

Reply:
I think all this is quite easy to verify...
```
$ tools/build.py --clean --logging=on --error-messages=on --debug
$ build/bin/jerry 
jerry> function fn(x) { switch (x) { case 1: { return } } }
function fn(x) { switch (x) { case 1: { return } } }
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
Script Error: SyntaxError: Primary expression expected. [line: 1, column: 48]
jerry> switch (x) { case 1: { return } }
switch (x) { case 1: { return } }
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
Script Error: SyntaxError: Primary expression expected. [line: 1, column: 31]
jerry> new Function("switch(1) { default: { return }}")
new Function("switch(1) { default: { return }}")
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
Script Error: SyntaxError: Primary expression expected. [line: 1, column: 31]
```

Reply:
The fix seems OK for the reported test case, but it would be great if this patch included the fix for the `new Function("switch(1) { default: { return }}")` case as well.

My suggestion for it:
``` diff
diff --git a/jerry-core/parser/js/js-parser.c b/jerry-core/parser/js/js-parser.c
index ec2198f4..d2c92d2e 100644
--- a/jerry-core/parser/js/js-parser.c
+++ b/jerry-core/parser/js/js-parser.c
@@ -2361,6 +2361,8 @@ parser_parse_source (const uint8_t *arg_list_p, /**< function argument list */
       context.source_end_p = source_p + source_size;
       context.line = 1;
       context.column = 1;
+      /* Note: When arg_list_p is not NULL, a function body is parsed */
+      context.status_flags |= PARSER_IS_FUNCTION;

       lexer_next_token (&context);
     }
```

@mjessome  Please take a look at it.

Reply:
Well, I checked more thoroughly the `new Function(...)` case and @mjessome solution seems enough and correct. I've missed [this](https://github.com/jerryscript-project/jerryscript/blob/master/jerry-core/parser/js/js-parser.c#L2286) part, where the context gets the `PARSER_IS_FUNCTION` status flag, so simply just ignore my suggestion.

Apart from that, it would be if the PR included the `new Function(...)` test case as well.